packages/devtools_app/lib/src/network/network_controller.dart

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:vm_service/vm_service.dart';
+
+import '../globals.dart';
+import '../ui/fake_flutter/fake_flutter.dart';
+import 'http_request_data.dart';
+
+class NetworkController {
+  /// Notifies that new HTTP requests have been processed.
+  ValueListenable get requestsNotifier => _httpRequestsNotifier;
+  final _httpRequestsNotifier = ValueNotifier<HttpRequests>(HttpRequests());
+
+  /// Notifies that the timeline is currently being recorded.
+  ValueListenable get recordingNotifier => _httpRecordingNotifier;
+  final _httpRecordingNotifier = ValueNotifier<bool>(false);
+
+  // The timeline timestamps are relative to when the VM started. This value is
+  // equal to `DateTime.now().microsecondsSinceEpoch - _profileStartMicros` when
+  // recording is started is used to calculate the correct wall-time for
+  // timeline events.
+  int _timelineMicrosOffset;
+  int _lastProfileRefreshMicros = 0;
+
+  void _processHttpTimelineEvents(Timeline timeline) {
+    final currentValues = _httpRequestsNotifier.value.requests;
+    final outstandingRequestsMap = _httpRequestsNotifier.value.outstanding;
+    final events = timeline.traceEvents;
+    final httpEventIds = <String>{};
+    // Perform initial pass to find the IDs for the HTTP timeline events.
+    for (final TimelineEvent event in events) {
+      final json = event.toJson();
+      final id = json['id'];
+      if (id == null) {
+        continue;
+      }
+      // The start event for HTTP contains a filter key which we can use.
+      // Note: only HTTP client requests are currently logged to the timeline.
+      if ((!json['args'].containsKey('filterKey') ||
+              json['args']['filterKey'] != 'HTTP/client') &&
+          !outstandingRequestsMap.containsKey(id)) {
+        continue;
+      }
+      httpEventIds.add(id);
+    }
+
+    // Group all HTTP timeline events with the same ID.
+    final Map<String, List<Map>> httpEvents = {};
+    for (final event in events) {
+      final json = event.toJson();
+      final id = json['id'];
+      if (id == null) {
+        continue;
+      }
+      if (httpEventIds.contains(id)) {
+        if (!httpEvents.containsKey(id)) {
+          httpEvents[id] = [];
+        }
+        httpEvents[id].add(json);
+      }
+    }
+
+    // Build our list of network requests from the collected events.
+    for (final request in httpEvents.entries) {
```

Comment:
I appreciate both your explanatory comments and your use of final.

Reply:
I appreciate your appreciation! ;-)